File: scripts/feature_detection.py
# -*- coding: utf-8 -*-
"""
feature_detection.py - Detect and report availability of optional features
"""

import logging
from typing import Dict, Optional
from dataclasses import dataclass

logger = logging.getLogger(__name__)


@dataclass
class FeatureStatus:
    """Status of an optional feature."""

    available: bool
    module_name: str
    error_message: Optional[str] = None
    installation_hint: Optional[str] = None


class FeatureDetector:
    """Detect availability of optional dependencies and features."""

    def __init__(self):
        self._cache: Dict[str, FeatureStatus] = {}
        self._detect_all()

    def _detect_all(self):
        """Detect all optional features at initialization."""
        # Voice features (STT/TTS)
        self._cache["voice_stt"] = self._detect_voice_stt()
        self._cache["voice_tts"] = self._detect_voice_tts()

        # PDF support
        self._cache["pdf"] = self._detect_pdf()

        # RAG features
        self._cache["rag"] = self._detect_rag()

        # Audio processing
        self._cache["audio"] = self._detect_audio()

        # Report availability
        available_features = [k for k, v in self._cache.items() if v.available]
        unavailable_features = [k for k, v in self._cache.items() if not v.available]

        logger.info(
            f"[FeatureDetector] Available features: {', '.join(available_features) if available_features else 'None'}"
        )
        if unavailable_features:
            logger.info(f"[FeatureDetector] Unavailable features: {', '.join(unavailable_features)}")

    def _detect_voice_stt(self) -> FeatureStatus:
        """Detect Speech-to-Text (Whisper) availability."""
        try:
            import torch
            import whisper

            return FeatureStatus(
                available=True,
                module_name="whisper",
            )
        except ImportError as e:
            missing_module = "torch" if "torch" in str(e) else "whisper"
            return FeatureStatus(
                available=False,
                module_name="whisper",
                error_message=f"Missing dependency: {missing_module}",
                installation_hint="Install with: pip install openai-whisper torch",
            )

    def _detect_voice_tts(self) -> FeatureStatus:
        """Detect Text-to-Speech availability."""
        try:
            import pyttsx3

            # Try to initialize TTS engine
            engine = pyttsx3.init()
            engine.stop()  # Clean up
            return FeatureStatus(
                available=True,
                module_name="pyttsx3",
            )
        except ImportError:
            return FeatureStatus(
                available=False,
                module_name="pyttsx3",
                error_message="Missing dependency: pyttsx3",
                installation_hint="Install with: pip install pyttsx3",
            )
        except Exception as e:
            return FeatureStatus(
                available=False,
                module_name="pyttsx3",
                error_message=f"TTS engine initialization failed: {e}",
                installation_hint="TTS engine may not be available on this system",
            )

    def _detect_pdf(self) -> FeatureStatus:
        """Detect PDF reading support."""
        try:
            import pypdf

            return FeatureStatus(
                available=True,
                module_name="pypdf",
            )
        except ImportError:
            return FeatureStatus(
                available=False,
                module_name="pypdf",
                error_message="Missing dependency: pypdf",
                installation_hint="Install with: pip install pypdf",
            )

    def _detect_rag(self) -> FeatureStatus:
        """Detect RAG dependencies (sentence-transformers, FAISS)."""
        try:
            import sentence_transformers
            import faiss

            return FeatureStatus(
                available=True,
                module_name="sentence_transformers, faiss",
            )
        except ImportError as e:
            missing_module = "sentence-transformers" if "sentence" in str(e) else "faiss"
            return FeatureStatus(
                available=False,
                module_name="sentence_transformers, faiss",
                error_message=f"Missing dependency: {missing_module}",
                installation_hint="Install with: pip install sentence-transformers faiss-cpu",
            )

    def _detect_audio(self) -> FeatureStatus:
        """Detect audio recording support."""
        try:
            import sounddevice as sd
            import scipy.io.wavfile

            return FeatureStatus(
                available=True,
                module_name="sounddevice, scipy",
            )
        except ImportError as e:
            missing_module = "sounddevice" if "sounddevice" in str(e) else "scipy"
            return FeatureStatus(
                available=False,
                module_name="sounddevice, scipy",
                error_message=f"Missing dependency: {missing_module}",
                installation_hint="Install with: pip install sounddevice scipy",
            )

    def is_available(self, feature: str) -> bool:
        """Check if a feature is available."""
        status = self._cache.get(feature)
        return status.available if status else False

    def get_status(self, feature: str) -> Optional[FeatureStatus]:
        """Get detailed status of a feature."""
        return self._cache.get(feature)

    def get_unavailable_reason(self, feature: str) -> str:
        """Get user-friendly message for why a feature is unavailable."""
        status = self._cache.get(feature)
        if not status:
            return f"Unknown feature: {feature}"

        if status.available:
            return f"{feature} is available"

        message = f"{feature.replace('_', ' ').title()} is not available"
        if status.error_message:
            message += f": {status.error_message}"
        if status.installation_hint:
            message += f"\n{status.installation_hint}"

        return message

    def get_all_unavailable(self) -> Dict[str, FeatureStatus]:
        """Get all unavailable features."""
        return {k: v for k, v in self._cache.items() if not v.available}


# Global instance
_detector = None


def get_feature_detector() -> FeatureDetector:
    """Get the global feature detector instance."""
    global _detector
    if _detector is None:
        _detector = FeatureDetector()
    return _detector


def is_feature_available(feature: str) -> bool:
    """Quick check if a feature is available."""
    return get_feature_detector().is_available(feature)

File: rust_output/src/feature_detection.rs
/// feature_detection::py - Detect and report availability of optional features

use anyhow::{Result, Context};
use serde::{Serialize, Deserialize};
use std::collections::HashMap;

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

pub static _DETECTOR: std::sync::LazyLock<Option<serde_json::Value>> = std::sync::LazyLock::new(|| None);

/// Status of an optional feature.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FeatureStatus {
    pub available: bool,
    pub module_name: String,
    pub error_message: Option<String>,
    pub installation_hint: Option<String>,
}

/// Detect availability of optional dependencies and features.
#[derive(Debug, Clone)]
pub struct FeatureDetector {
    pub _cache: HashMap<String, FeatureStatus>,
}

impl FeatureDetector {
    pub fn new() -> Self {
        Self {
            _cache: HashMap::new(),
        }
    }
    /// Detect all optional features at initialization.
    pub fn _detect_all(&mut self) -> () {
        // Detect all optional features at initialization.
        self._cache["voice_stt".to_string()] = self._detect_voice_stt();
        self._cache["voice_tts".to_string()] = self._detect_voice_tts();
        self._cache["pdf".to_string()] = self._detect_pdf();
        self._cache["rag".to_string()] = self._detect_rag();
        self._cache["audio".to_string()] = self._detect_audio();
        let mut available_features = self._cache.iter().iter().filter(|(k, v)| v.available).map(|(k, v)| k).collect::<Vec<_>>();
        let mut unavailable_features = self._cache.iter().iter().filter(|(k, v)| !v.available).map(|(k, v)| k).collect::<Vec<_>>();
        logger.info(format!("[FeatureDetector] Available features: {}", if available_features { available_features.join(&", ".to_string()) } else { "None".to_string() }));
        if unavailable_features {
            logger.info(format!("[FeatureDetector] Unavailable features: {}", unavailable_features.join(&", ".to_string())));
        }
    }
    /// Detect Speech-to-Text (Whisper) availability.
    pub fn _detect_voice_stt(&self) -> Result<FeatureStatus> {
        // Detect Speech-to-Text (Whisper) availability.
        // try:
        {
            // TODO: import torch
            // TODO: import whisper
            FeatureStatus(/* available= */ true, /* module_name= */ "whisper".to_string())
        }
        // except ImportError as e:
    }
    /// Detect Text-to-Speech availability.
    pub fn _detect_voice_tts(&self) -> Result<FeatureStatus> {
        // Detect Text-to-Speech availability.
        // try:
        {
            // TODO: import pyttsx3
            let mut engine = pyttsx3.init();
            engine::stop();
            FeatureStatus(/* available= */ true, /* module_name= */ "pyttsx3".to_string())
        }
        // except ImportError as _e:
        // except Exception as e:
    }
    /// Detect PDF reading support.
    pub fn _detect_pdf(&self) -> Result<FeatureStatus> {
        // Detect PDF reading support.
        // try:
        {
            // TODO: import pypdf
            FeatureStatus(/* available= */ true, /* module_name= */ "pypdf".to_string())
        }
        // except ImportError as _e:
    }
    /// Detect RAG dependencies (sentence-transformers, FAISS).
    pub fn _detect_rag(&self) -> Result<FeatureStatus> {
        // Detect RAG dependencies (sentence-transformers, FAISS).
        // try:
        {
            // TODO: import sentence_transformers
            // TODO: import faiss
            FeatureStatus(/* available= */ true, /* module_name= */ "sentence_transformers, faiss".to_string())
        }
        // except ImportError as e:
    }
    /// Detect audio recording support.
    pub fn _detect_audio(&self) -> Result<FeatureStatus> {
        // Detect audio recording support.
        // try:
        {
            // TODO: import sounddevice as sd
            // TODO: import scipy.io.wavfile
            FeatureStatus(/* available= */ true, /* module_name= */ "sounddevice, scipy".to_string())
        }
        // except ImportError as e:
    }
    /// Check if a feature is available.
    pub fn is_available(&mut self, feature: String) -> bool {
        // Check if a feature is available.
        let mut status = self._cache.get(&feature).cloned();
        if status { status.available } else { false }
    }
    /// Get detailed status of a feature.
    pub fn get_status(&self, feature: String) -> Option<FeatureStatus> {
        // Get detailed status of a feature.
        self._cache.get(&feature).cloned()
    }
    /// Get user-friendly message for why a feature is unavailable.
    pub fn get_unavailable_reason(&mut self, feature: String) -> String {
        // Get user-friendly message for why a feature is unavailable.
        let mut status = self._cache.get(&feature).cloned();
        if !status {
            format!("Unknown feature: {}", feature)
        }
        if status.available {
            format!("{} is available", feature)
        }
        let mut message = format!("{} is not available", /* title */ feature.replace(&*"_".to_string(), &*" ".to_string()).to_string());
        if status.error_message {
            message += format!(": {}", status.error_message);
        }
        if status.installation_hint {
            message += format!("\n{}", status.installation_hint);
        }
        message
    }
    /// Get all unavailable features.
    pub fn get_all_unavailable(&self) -> HashMap<String, FeatureStatus> {
        // Get all unavailable features.
        self._cache.iter().iter().filter(|(k, v)| !v.available).map(|(k, v)| (k, v)).collect::<HashMap<_, _>>()
    }
}

/// Get the global feature detector instance.
pub fn get_feature_detector() -> FeatureDetector {
    // Get the global feature detector instance.
    // global/nonlocal _detector
    if _detector.is_none() {
        let mut _detector = FeatureDetector();
    }
    _detector
}

/// Quick check if a feature is available.
pub fn is_feature_available(feature: String) -> bool {
    // Quick check if a feature is available.
    get_feature_detector().is_available(feature)
}
